python/ql/src/experimental/Security/CWE-090/LDAPInjection.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @name Python LDAP Injection
+ * @description Python LDAP Injection through search filter
+ * @kind path-problem
+ * @problem.severity error
+ * @id python/ldap-injection
+ * @tags experimental	
+ *       security	
+ *       external/cwe/cwe-090
+ */
+
+// Determine precision above
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.TaintTrackingPublic
+import DataFlow::PathGraph
+
+class InitializeSink extends DataFlow::Node {
+  InitializeSink() {
+    exists(SsaVariable initVar, CallNode searchCall |
+      // get variable whose value equals a call to ldap.initialize
+      initVar.getDefinition().getImmediateDominator() = Value::named("ldap.initialize").getACall() and
+      // get the Call in which the previous variable is used
+      initVar.getAUse().getNode() = searchCall.getNode().getFunc().(Attribute).getObject() and
+      // restrict that call's attribute (something.this) to match %search%
+      searchCall.getNode().getFunc().(Attribute).getName().matches("%search%") and
+      // set the third argument (search_filter) as sink
+      this.asExpr() = searchCall.getArg(2).getNode()
+      // set the first argument (DN) as sink
+      // or this.asExpr() = searchCall.getArg(0) // Should this be set?
+    )
+  }
+}
```

Comment:
Since I haven't been able to achieve `ldap_connection.search_s` this way (although looking at a test's `AST` shows the same result as `Regex Injection` query):

```codeql
class InitializeSink extends DataFlow::Node {
  InitializeSink() {
    this =
      API::moduleImport("ldap")
          .getMember("initialize")
          .getMember("search_s")
          .getACall()
          .(DataFlow::CallCfgNode)
          .getArg(2)
  }
}
```

I've tried this way of taking `ldap.initialize` as a `CallCfgNode` and `.search_s` as a `AttrRead` but `searchCall.(DataFlow::CallCfgNode)` doesn't seem to work for this.

```codeql
class InitializeSink extends DataFlow::Node {
  InitializeSink() {
    exists(DataFlow::CallCfgNode initVar, DataFlow::AttrRead searchCall |
      initVar = API::moduleImport("ldap").getMember("initialize").getACall() and
      searchCall.getObject().getALocalSource() = initVar and
      searchCall.getAttributeName().matches("%search%") and
      this = searchCall.(DataFlow::CallCfgNode).getArg(2)
    )
  }
}
```

Therefore I've tried using a third variable being set to `searchCall` by calling both `.asCfgNode()` but doesn't work either.

```codeql
class InitializeSink extends DataFlow::Node {
  InitializeSink() {
    exists(
      DataFlow::CallCfgNode initVar, DataFlow::AttrRead searchCall, DataFlow::CallCfgNode searchArg
    |
      initVar = API::moduleImport("ldap").getMember("initialize").getACall() and
      searchCall.getObject().getALocalSource() = initVar and
      searchCall.getAttributeName().matches("%search%") and
      searchArg.asCfgNode() = searchCall.asCfgNode() and
      this = searchArg.getArg(2)
    )
  }
}
```

Reply:
Got it :D